File: hot/剑指offer/32 - III.从上到下打印二叉树 III/32 - III.从上到下打印二叉树 III.py
from typing import List

from util.leetcode_type import TreeNode

"""
奇数正序，偶数倒序
每次外层循环，设置临时数组存放，在循环结束是判断层级奇偶来决定临时数组的结果倒序还是正序
"""


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        result, current, level = [], [root], 1
        while current:
            next_level, vals = [], []
            for node in current:
                vals.append(node.val)
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            if level % 2:
                result.append(vals)
            else:
                result.append(vals[::-1])
            current = next_level
            level += 1
        return result

File: hot/剑指offer/32 - III.从上到下打印二叉树 III/32 - III.从上到下打印二叉树 III.go
package hot

import "algorithms-code-note/util/structures"

type TreeNode = structures.TreeNode

func levelOrder(root *TreeNode) [][]int {
	result := make([][]int, 0)
	if root == nil {
		return result
	}
	stk := []*TreeNode{root}
	level := 1
	for len(stk) != 0 {
		length := len(stk)
		vals := make([]int, length)
		tmp := []*TreeNode{}
		for i := 0; i < length; i++ {
			if stk[i] == nil{
				continue
			}
			if level%2 != 0 {
				vals[i] = stk[i].Val
			} else {
				vals[length-1-i] = stk[i].Val
			}
			if stk[i].Left != nil {
				tmp = append(tmp, stk[i].Left)
			}
			if stk[i].Right != nil {
				tmp = append(tmp, stk[i].Right)
			}
		}
		stk = tmp
		result = append(result, vals)
		level += 1
	}
	return result
}
